### ISSUE TITLE
[Bug] Preview overlays on top of the prompt when using `split`

### ISSUE THREAD

**Issue by asmodeus812** [ibhagwan/fzf-lua#779]
Hi, i recently started noticing (since a few days) that for a horizontal preview the preview list overlaps the first row which contains the search prompt. This now is happening when i use the split option in winopts. Oddly enough it is working in terminals but in neovide it is obscured, only the prompt line, by the preview window i think. So the issue is only reproducible in neovide. As it was working fine up until very recently, i guess there could be something related to the recent border commits, but i am not sure. (Note: that trying to set the border to "none" just completely locks nvim, if i try to open any fzf picker, so i was unable to test with that)

```lua
       local default_cmd = "belowright new | wincmd J | resize 40" 
        winopts = {
            split = default_cmd,
            height = 0.85,
            width = 0.85,
            row = 0.35,
            col = 0.35,
            preview = {
                delay = 100,
                title = true,
                wrap = "nowrap",
                hidden = "nohidden",
                border = "border",
                default = "builtin",
                layout = "vertical",
                vertical = "up:55%",
                horizontal = "right:55%",
                title_align = "center",
            },
        },

```

![image](https://github.com/ibhagwan/fzf-lua/assets/15955811/b91e9812-af37-4e13-b6c0-e84873ad7b9e)

**Comment by ibhagwan** [comment]
@asmodeus812, that's because the code does an estimated calculation of the positioning and then attempts to overlay the neovim preview window on top of the terminal window, the calc isn't perfect and I'm assuming using neovide modifies the total rows which causes this affect, you can probably play with `height` and `preview.vertical` to get to a layout that works.

I'm not sure how to solve this or if I can get to a perfect alignment in all cases, will need to be investigated.

**Comment by asmodeus812** [comment]
@ibhagwan okay yes, so i tried it on multiple machines, but have to adjust either the font size or the width percentage to force the prompt to show up, for example what seems to work on one resolution fails to work on another and it can be finicky. Should i assume this is neovide specific issue, or should i try to test with a few terminals for a few days too ? Like alacritty is what i usually use. 
For the border property what i meant is that i set "none" to the winopts.preview.border to taht, instead of "border" i wanted to see if the border is what obscures it but as i said "none" is not working, the entire nvim instance locks up and have to restart the process completely.

**Comment by ibhagwan** [comment]
> For the border property what i meant is that i set "none" to the winopts.preview.border to taht, instead of "border" i wanted to see if the border is what obscures it but as i said "none" is not working, the entire nvim instance locks up and have to restart the process completely.

In regards to this, it wouldn't make any difference as the preview border option only applies to fzf native previewers such as `bat`, you're using the "builtin" neovim previewer which draws a neovim window ontop of the fzf window.

That said, although not a solution but a workaround, using `bat` previewer will solve this issue as fzf will be responsible for both the content and preview (in the terminal) so there won't be any alignment issues, you can try with `:FzfLua files previewer=bat`.

As for the `none` option, that doesn't work beccause it's not a valid `fzf` preview window option, if you want to remove the fzf preview border set it to `border-none`, from `man fzf` (but again, you will not see the difference unless you're using `bat` for preview):
```
              * To change the style of the border of the preview window,
              specify one of the options for --border with border- prefix.
              e.g. border-rounded (border with rounded edges, default),
              border-sharp (border with sharp edges), border-left, border-
              none, etc.
```

**Comment by asmodeus812** [comment]
@ibhagwan hey, i don't want to open a new issue, and this one is vaguely related to the one i am using now. Using the winopts.split option, what i want to to is restore the view when the picker is closed, see below.

```lua

    -- this has to happen before the picker is opened, problem is we 
    -- don't have a way to intercept the picker before its view gets shown to the user
    -- imo we should think about some user events to have more control over the picker states / stages ?
    local state = vim.fn.winrestcmd()
    vim.t._picker_winstate = state

   -- restore when closed
       vim.api.nvim_create_autocmd({ "TermClose" }, {
        pattern = "term://*",
        callback = function()
            if vim.bo.ft == "fzf" and vim.t._picker_winstate then
                pcall(vim.cmd, "exec t:_picker_winstate")
                vim.t._picker_winstate = nil
            end
        end,
        desc = "Restoring window state after picker closed",
    })
```

**Comment by ibhagwan** [comment]
Hi @asmodeus812, you didn’t respond to my other questions in regards to neovide split overlay issue, is the issue no longer relevant?

As for your question, I’m not sure if you’re asking me if that’s the correct way to do it using the `TermClose` event? Did you try it? Does it not work?

You can also try defining a `fn_post_fzf` function, I’m not sure if it can currently defined in the global setup but try it in the direct function opts and if it solves your issue I can also take it into account when defined globally:
https://github.com/ibhagwan/fzf-lua/blob/1bba731df46feb1751dca1632268aae41ed5ac15/lua/fzf-lua/core.lua#L328

**Comment by asmodeus812** [comment]
Hi @ibhagwan,

Yes on the question of the initial issue, it is happening depending on the font size or font type, for example on macos with external  4k monitor and Jetbrains NF, with a font of 18 it works fine, but on the native macbook screen its obscured. It is very fiddly

The question i have at the moment is not related to how to use TermClose no, the problem is very simple. When you open fzf in a non-floating window using the `split` option to do something like that `belowright new | wincmd J` the other windows you have before you open the fzf picker, are auto-resized by vim proportionally (which is fine), once you close the fzf picker however, the ones that were re-sized remain resized (in my case since i use wincmd J, the windows will move to make space at the bottom for the picker window, however, when you close the fzf picker the now empty space is filled up by the closest horiz spit window, above, which is re-sized to fit that space left over by the closed fzf split window)  what i want is to restore their sizes as they were before the fzf window was opened (see below)

The post_fzf function might work, but there are places such as `workspace_symbols` where the post function is getting re-assigned, so there is no way for a user to pass his own post_fn in some places, unless handled correctly. There is also pre_fzf hook i saw somewhere. 

Another caveat is that those have to be passed in to all pickers, would be great if the hooks can be picked up from the current picker opts, or if not present from the user config to fzf_lua. 

Initial layout
![image](https://github.com/ibhagwan/fzf-lua/assets/15955811/a0bda6f0-486c-4961-b47c-c57d26ff449d)

After picker opens
![image](https://github.com/ibhagwan/fzf-lua/assets/15955811/97cb0ed2-b0a0-4ea7-8a2f-d0be0bdabea5)

After picker closes
![image](https://github.com/ibhagwan/fzf-lua/assets/15955811/2fd856b8-f3cf-4fe7-b8c6-d8f49e3a76c7)

**Comment by ibhagwan** [comment]
@asmodeus812, can you update to https://github.com/ibhagwan/fzf-lua/commit/b1aaac28eb430a245598f2b5580eb6bee4d70efe  and try the `winopts.on_close` event?

Also, if you’d like me to solve the OP please post the responses to the steps I mentioned in the above post.

**Comment by asmodeus812** [comment]
> ```lua
> lua print(vim.o.lines, vim.o.columns)
> ```

Hi @ibhagwan haven't had time to do that on the pc where that was reproducible, we can close this one as i can't find the time to debug the issue either. But i have found another issue which broke the split command usage for me, will create issue for it.
